src/MicrosoftTeams.ts

```
@@ -417,25 +442,132 @@ namespace microsoftTeams
+                top += (currentWindow.outerHeight / 2) - (height / 2);
+
+                // Open the window with a desired set of standard browser features
+                childWindow = currentWindow.open(link.href, "_blank", "toolbar=no, location=yes, status=no, menubar=no, top=" + top + ", left=" + left + ", width=" + width + ", height=" + height);
+            }
+
+            if (!childWindow)
+            {
+                throw new Error("Failed to open authentication window.");
+            }
+
+            // Start monitoring the authentication window so that we can detect if it gets closed before the flow completes
+            startAuthenticationWindowMonitor();
+        }
+
+        function stopAuthenticationWindowMonitor(): void
+        {
+            if (authWindowMonitor)
+            {
+                clearInterval(authWindowMonitor);
+                authWindowMonitor = 0;
+            }
+
+            handlers["initialize"] = null;
+        }
+
+        function startAuthenticationWindowMonitor(): void
+        {
+            // Stop the previous window monitor if there is one running
+            stopAuthenticationWindowMonitor();
+
+            // Create an interval loop that:
+            // - Notifies the caller of failure if it detects that the authentication window was closed
+            // - Keeps pinging the authentication window until it establishes contact
+            authWindowMonitor = setInterval(() =>
+            {
+                if (!childWindow || childWindow.closed)
+                {
+                    handleFailure("CancelledByUser");
                 }
-                else
+                else if (!childOrigin)
                 {
-                    authenticateParameters.failureCallback(response);
+                    try
+                    {
+                        childOrigin = "*";
+                        sendMessageRequest(childWindow, "ping");
+                    }
```

I've seen this call throw before but I can't seem to reproduce it now. I mostly added the try...finally because I wanted to make sure that we always clean up the childWindow and childOrigin regardless of whether closing the window was successful or not. At this point I would prefer to let the exception bubble out so that I can make sure this call isn't failing in some unexpected cases. I might just completely remove the try...finally at some point.